src/algebra/module/torsion.lean

```diff
@@ -223,77 +190,122 @@ eq_bot_iff.mpr (λ _ h, by { rw [mem_torsion_by_iff, one_smul] at h, exact h })
 by { rw [eq_bot_iff, ← torsion_by_one, ← torsion_by_singleton_eq],
   exact torsion_by_set_le_torsion_by_set_of_subset (λ _ _, trivial) }
 
+end submodule
+open submodule
+namespace module
+
+@[simp] lemma is_torsion_by_singleton_iff : is_torsion_by_set R M {a} ↔ is_torsion_by R M a :=
+begin
+  refine ⟨λ h x, @h _ ⟨_, set.mem_singleton _⟩, λ h x, _⟩,
+  rintro ⟨b, rfl : b = a⟩, exact @h _
+end
+
+lemma is_torsion_by_set_iff_torsion_by_set_eq_top :
+  is_torsion_by_set R M s ↔ submodule.torsion_by_set R M s = ⊤ :=
+⟨λ h, eq_top_iff.mpr (λ _ _, (mem_torsion_by_set_iff _ _).mpr $ @h _),
+  λ h x, by { rw [← mem_torsion_by_set_iff, h], trivial }⟩
+
+/-- A `a`-torsion module is a module whose `a`-torsion submodule is the full space. -/
+lemma is_torsion_by_iff_torsion_by_eq_top : is_torsion_by R M a ↔ torsion_by R M a = ⊤ :=
+by rw [← torsion_by_singleton_eq, ← is_torsion_by_singleton_iff,
+  is_torsion_by_set_iff_torsion_by_set_eq_top]
+
+lemma is_torsion_by_set_iff_is_torsion_by_span :
+  is_torsion_by_set R M s ↔ is_torsion_by_set R M (ideal.span s) :=
+by rw [is_torsion_by_set_iff_torsion_by_set_eq_top, is_torsion_by_set_iff_torsion_by_set_eq_top,
+  torsion_by_set_eq_torsion_by_span]
+
+lemma is_torsion_by_span_singleton_iff : is_torsion_by_set R M (R ∙ a) ↔ is_torsion_by R M a :=
+((is_torsion_by_set_iff_is_torsion_by_span _).symm.trans $ is_torsion_by_singleton_iff _)
+
+end module
+namespace submodule
+open module
+
+lemma torsion_by_set_is_torsion_by_set : is_torsion_by_set R (torsion_by_set R M s) s :=
+λ ⟨x, hx⟩ a, subtype.ext $ (mem_torsion_by_set_iff _ _).mp hx a
+
+/-- The `a`-torsion submodule is a `a`-torsion module. -/
+lemma torsion_by_is_torsion_by : is_torsion_by R (torsion_by R M a) a := λ _, smul_torsion_by _ _
+
+@[simp] lemma torsion_by_torsion_by_eq_top : torsion_by R (torsion_by R M a) a = ⊤ :=
+(is_torsion_by_iff_torsion_by_eq_top a).mp $ torsion_by_is_torsion_by a
+@[simp] lemma torsion_by_set_torsion_by_set_eq_top :
+  torsion_by_set R (torsion_by_set R M s) s = ⊤ :=
+(is_torsion_by_set_iff_torsion_by_set_eq_top s).mp $ torsion_by_set_is_torsion_by_set s
+
+variables (R M)
+lemma torsion_gc : @galois_connection (submodule R M) (ideal R)ᵒᵈ _ _
+  annihilator (λ I, torsion_by_set R M $ I.of_dual) :=
+λ A I, ⟨λ h x hx, (mem_torsion_by_set_iff _ _).mpr $ λ ⟨a, ha⟩, mem_annihilator.mp (h ha) x hx,
+  λ h a ha, mem_annihilator.mpr $ λ x hx, (mem_torsion_by_set_iff _ _).mp (h hx) ⟨a, ha⟩⟩
+
+variables {R M}
 section coprime
 open_locale big_operators
-open dfinsupp
-variables {ι : Type*} {p : ι → R} {S : finset ι} (hp : pairwise (is_coprime on λ s : S, p s))
+variables {ι : Type*} {p : ι → ideal R} {S : finset ι}
+variables (hp : (S : set ι).pairwise $ λ i j, p i ⊔ p j = ⊤)
 include hp
 
-lemma supr_torsion_by_eq_torsion_by_prod :
-  (⨆ i : S, torsion_by R M (p i)) = torsion_by R M (∏ i in S, p i) :=
+lemma supr_torsion_by_ideal_eq_torsion_by_infi :
```

Comment:
I don't understand what you are doing here. What happened to `supr_torsion_by_eq_torsion_by_prod`?

Reply:
This is a generalisation, when the ideals `p i` are principal you get back the former `supr_torsion_by_eq_torsion_by_prod` lemma. Maybe I should keep a lemma for this specific case explicitly ?

Reply:
Yes, I think it is a good idea.

Reply:
Done